bricks/system/syscalls.h

```diff
@@ -0,0 +1,214 @@
+/*******************************************************************************
+The MIT License (MIT)
+
+Copyright (c) 2018 Dmitry "Dima" Korolev <dmitry.korolev@gmail.com>
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+#ifndef BRICKS_SYSTEM_SYSCALLS_H
+#define BRICKS_SYSTEM_SYSCALLS_H
+
+#include "../../port.h"
+
+#include "../exception.h"
+#include "../file/file.h"
+#include "../strings/util.h"
+
+#include <dlfcn.h>
+
+#include <cstdio>
+#include <vector>
+
+namespace current {
+namespace bricks {
+namespace system {
+
+struct SystemException : Exception {
+  using Exception::Exception;
+};
+
+struct PopenCallException final : SystemException {
+  using SystemException::SystemException;
+};
+
+struct SystemCallException final : SystemException {
+  using SystemException::SystemException;
+};
+
+class InputTextPipe final {
+ private:
+  const std::string command_;
+  std::vector<char> buffer_;
+  FILE* f_;
+
+ public:
+  template <typename S>
+  explicit InputTextPipe(S&& command, size_t max_line_length = 1024 * 1024)
```

Comment:
Using large buffers is usually a bad idea.
Typical buffer size for various file/network operations is 1-16K

Reply:
True, but this is not just the buffer size, it's the maximum line length for `fread`; the output would make little sense if at least one line exceeds it.

Changed to 16K in a84c2d39 regardless. :)

Reply:
The thing is that in this code it's both, max line length **and** the buffer size.
Proper buffering implementation may be added later though.

Reply:
Yes, definitely not now.